### PR TITLE
Added Same() and NotSame() assertions.

### PR DESCRIPTION
**Pull request by Monnoroch**
They are similar to Exactly() and NotEqual(), but test equality by reference instead of by value.

### CODE REVIEW

File: assert/assertions.go [hunk 1]
```diff
@@ -253,6 +253,51 @@ func IsType(t TestingT, expectedType interface{}, object interface{}, msgAndArgs
 	return true
 }
 
+// Equal asserts that two objects are actually the same one.
```
Comment (exarkun) [inline assert/assertions.go]: Should say "Same asserts ..."

File: assert/assertions.go [hunk 2]
```diff
@@ -253,6 +253,51 @@ func IsType(t TestingT, expectedType interface{}, object interface{}, msgAndArgs
...
+
+	if expected != actual {
+		return Fail(t, fmt.Sprintf("Not the same: %#v (expected)\n"+
+			"        != %#v (actual)%s", expected, actual), msgAndArgs...)
+	}
+
+	return true
+}
+
+// Equal asserts that two objects are actually the same one.
```
Comment (exarkun) [inline assert/assertions.go]: Similar comment here.

File: assert/assertions_test.go [hunk 3]
```diff
@@ -94,6 +94,11 @@ func (a *AssertionTesterConformingObject) TestMethod() {
 type AssertionTesterNonConformingObject struct {
 }
 
+// AssertionTesterConformingObject is an object that conforms to the AssertionTesterInterface interface
```
Comment (exarkun) [inline assert/assertions_test.go]: "samenessTestHelper is ..."

File: require/require.go [hunk 4]
```diff
@@ -397,6 +397,18 @@ func NotRegexp(t TestingT, rx interface{}, str interface{}, msgAndArgs ...interf
 }
 
 
+// Equal asserts that two objects are actually the same one.
```
Comment (exarkun) [inline require/require.go]: "NotSame ..."

File: require/require.go [hunk 5]
```diff
@@ -432,6 +444,19 @@ func Regexp(t TestingT, rx interface{}, str interface{}, msgAndArgs ...interface
...
```
Comment (exarkun) [inline require/require.go]: "Same ..."

File: require/require_forward.go [hunk 6]
```diff
@@ -333,6 +333,16 @@ func (a *Assertions) NotRegexp(rx interface{}, str interface{}, msgAndArgs ...in
 }
 
 
+// Equal asserts that two objects are actually the same one.
```
Comment (exarkun) [inline require/require_forward.go]: "NotSame ..."

File: require/require_forward.go [hunk 7]
```diff
@@ -362,6 +372,17 @@ func (a *Assertions) Regexp(rx interface{}, str interface{}, msgAndArgs ...inter
...
```
Comment (exarkun) [inline require/require_forward.go]: "Same ..."

File: assert/assertion_forward.go [hunk 8]
```diff
@@ -332,6 +332,16 @@ func (a *Assertions) NotRegexp(rx interface{}, str interface{}, msgAndArgs ...in
 }
 
 
+// Equal asserts that two objects are actually the same one.
```
Comment (exarkun) [inline assert/assertion_forward.go]: "NotSame ..."

File: assert/assertion_forward.go [hunk 9]
```diff
@@ -361,6 +371,17 @@ func (a *Assertions) Regexp(rx interface{}, str interface{}, msgAndArgs ...inter
...
```
Comment (exarkun) [inline assert/assertion_forward.go]: "Same ..."

File: assert/assertions.go (hunk 1 above)
Comment (Monnoroch) [inline assert/assertions.go]: Fixed.

File: assert/assertions.go (hunk 2 above)
Comment (Monnoroch) [inline assert/assertions.go]: Fixed.

File: assert/assertions_test.go (hunk 3 above)
Comment (Monnoroch) [inline assert/assertions_test.go]: Fixed.

File: require/require.go (hunk 4 above)
Comment (Monnoroch) [inline require/require.go]: Fixed.

File: require/require.go (hunk 5 above)
Comment (Monnoroch) [inline require/require.go]: Fixed.

File: require/require_forward.go (hunk 6 above)
Comment (Monnoroch) [inline require/require_forward.go]: Fixed.

File: require/require_forward.go (hunk 7 above)
Comment (Monnoroch) [inline require/require_forward.go]: Fixed.

File: assert/assertion_forward.go (hunk 8 above)
Comment (Monnoroch) [inline assert/assertion_forward.go]: Fixed.

File: assert/assertion_forward.go (hunk 9 above)
Comment (Monnoroch) [inline assert/assertion_forward.go]: Fixed.

**Comment by Monnoroch** [comment]
Thanks!
Fixed all the comments.

**Comment by ernesto-jimenez** [comment]
@Monnoroch sorry it took so long to reply!

Would it be better to do the following?

```go
assert.True(t, expected == actual)
// or
assert.True(t, expected == actual, "%#v should be %#v", expected, actual)
```

`assert.True` would have the added benefit that a type mismatch would be detected by the compiler instead of on runtime using reflection.

**Comment by Monnoroch** [comment]
Fair enough. The only downside is writing the error message yourself.

**Pull request closed without merging**